### PR TITLE
oneAPI kmeans

### PR DESCRIPTION
**Pull request by mchernov-intel**
moved from: #680

### CODE REVIEW

File: cpp/oneapi/dal/algo/kmeans/backend/cpu/train_kernel_lloyd_dense.cpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        interop::call_daal_kernel<Float, daal_kmeans_init_plus_plus_dense_kernel_t>(ctx,
+                                                                                    init_len_input,
+                                                                                    init_input,
+                                                                                    init_len_output,
+                                                                                    init_output,
+                                                                                    &par,
+                                                                                    *(par.engine));
+
+        daal::data_management::NumericTablePtr daal_centroids_ptr(daal_centroids);
+        new_initial_centroids = interop::convert_from_daal_homogen_table<Float>(daal_centroids_ptr);
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/algo/kmeans/backend/cpu/train_kernel_lloyd_dense.cpp:84]:
```suggestion
        new_initial_centroids = interop::convert_from_daal_homogen_table<Float>(daal_centroids);
```

File: cpp/oneapi/dal/algo/kmeans/backend/cpu/train_kernel_lloyd_dense.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            daal_centroids.get()
+        };
+
+        interop::call_daal_kernel<Float, daal_kmeans_init_plus_plus_dense_kernel_t>(ctx,
+                                                                                    init_len_input,
+                                                                                    init_input,
+                                                                                    init_len_output,
+                                                                                    init_output,
+                                                                                    &par,
+                                                                                    *(par.engine));
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/algo/kmeans/backend/cpu/train_kernel_lloyd_dense.cpp:81]:
```suggestion
        const size_t init_len_input                                     = 1;
        daal::data_management::NumericTable* init_input[init_len_input] = { daal_data.get() };

        auto daal_centroids =
            interop::allocate_daal_homogen_table<Float>(cluster_count, column_count);
        const size_t init_len_output                                      = 1;
        daal::data_management::NumericTable* init_output[init_len_output] = {
            daal_centroids.get()
        };

        interop::call_daal_kernel<Float, daal_kmeans_init_plus_plus_dense_kernel_t>(ctx,
                                                                                    1,
                                                                                    daal_data.get(),
                                                                                    1,
                                                                                    daal_centroids.get(),
                                                                                    &par,
                                                                                    *(par.engine));
```

**Comment by mchernov-intel** [comment]
@PetrovKP @rlnx guys, please, review

**Comment by PetrovKP** [comment]
@mchernov-intel Please, see https://github.com/oneapi-src/oneDAL/pull/680. Most comments are not addressed. 😩

File: examples/oneapi/cpp/source/kmeans_init/kmeans_init_default_dense.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+#include "oneapi/dal/algo/kmeans_init.hpp"
+#include "example_util/utils.hpp"
+
+using namespace oneapi;
+
+int main(int argc, char const* argv[]) {
+    constexpr std::int64_t row_count    = 8;
+    constexpr std::int64_t column_count = 7;
+
+    const size_t nClusters = 2;
```
Comment (rlnx) [inline examples/oneapi/cpp/source/kmeans_init/kmeans_init_default_dense.cpp:29]: Code style: `cluster_count`

Comment (rlnx) [reply examples/oneapi/cpp/source/kmeans_init/kmeans_init_default_dense.cpp:29]: `std::int64_t`

File: cpp/oneapi/dal/algo/kmeans/detail/infer_ops.hpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+template <typename Descriptor>
+struct infer_ops {
+    using float_t           = typename Descriptor::float_t;
+    using method_t          = method::by_default;
+    using input_t           = infer_input;
+    using result_t          = infer_result;
+    using descriptor_base_t = descriptor_base;
+
+    void validate(const Descriptor& params, const infer_input& input) const {}
```
Comment (PetrovKP) [inline cpp/oneapi/dal/algo/kmeans/detail/infer_ops.hpp:36]: Why validate is empty?

File: cpp/oneapi/dal/algo/kmeans/detail/train_ops.hpp [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    using float_t           = typename Descriptor::float_t;
+    using method_t          = typename Descriptor::method_t;
+    using input_t           = train_input;
+    using result_t          = train_result;
+    using descriptor_base_t = descriptor_base;
+
+    void validate(const Descriptor& params, const train_input& input) const {
+        if (!(input.get_data().has_data())) {
+            throw domain_error("Input data is empty");
+        }
```
Comment (PetrovKP) [inline cpp/oneapi/dal/algo/kmeans/detail/train_ops.hpp:40]: You miss checks of parameter

cluster_num >= 0
iter_num >=0
acc_threashold >=0

File: examples/oneapi/cpp/source/kmeans_init/kmeans_init_default_dense.cpp [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+                           4.f, 2.f,  0.f, -4.f, 0.f,  3.f,  -8.f,
+                           2.f,  5.f, 5.f,  -6.f, 3.f, 0.f, -9.f,
+                           3.f, 1.f,  -3.f, 3.f,  5.f,  1.f,  7.f };
+
+    const auto data_table = dal::homogen_table{ row_count, column_count, data };
+
+    const auto kmeans_init_desc = dal::kmeans_init::descriptor<>()
+        .set_cluster_count(cluster_count);
+
+    const auto result = dal::train(kmeans_init_desc, data_table);
```
Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans_init/kmeans_init_default_dense.cpp:45]: @rlnx  we real want have a `train` method for `kmearn_init` algorithm?

File: examples/oneapi/cpp/source/kmeans/kmeans_lloyd_dense_batch.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+                                     1.f, 2.f, 5.f, 2.f, 9.f, 3.f, 2.f};
+
+  const auto data_table = dal::homogen_table{row_count, column_count, data};
+  const auto initial_centroids_table =
+      dal::homogen_table{1, column_count, initial_centroids};
+
+  const auto kmeans_desc = dal::kmeans::descriptor<>()
+                               .set_cluster_count(2)
+                               .set_max_iteration_count(5)
+                               .set_accuracy_threshold(1.0);
```
Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans/kmeans_lloyd_dense_batch.cpp:43]: Why do you have only 5 iterations and a threshold of more than 1? Judging by the objective function, you do not solve the problem well. Maybe tighten the parameters?

File: examples/oneapi/cpp/source/kmeans/kmeans_lloyd_dense_batch.cpp [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+                        1.f, -3.f, 3.f, 5.f, 1.f, 7.f};
+
+  const float initial_centroids[] = {1.f, 2.f, 3.f, 4.f, 5.f, 6.f, -5.f,
+                                     1.f, 2.f, 5.f, 2.f, 9.f, 3.f, 2.f};
+
+  const auto data_table = dal::homogen_table{row_count, column_count, data};
+  const auto initial_centroids_table =
+      dal::homogen_table{1, column_count, initial_centroids};
+
+  const auto kmeans_desc = dal::kmeans::descriptor<>()
```
Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans/kmeans_lloyd_dense_batch.cpp:40]:
```suggestion
  const auto kmeans_desc = dal::kmeans::descriptor{}
```

File: cpp/oneapi/dal/algo/pca/detail/train_ops.hpp [hunk 9]
```diff
@@ -33,7 +34,14 @@ struct train_ops {
...
     using descriptor_base_t = descriptor_base;
 
-    void validate(const Descriptor& params, const train_input& input) const {}
+    void validate(const Descriptor& params, const train_input& input) const {
+        if (!(input.get_data().has_data())) {
+            throw domain_error("Input data is empty.");
+        }
+        if (input.get_data().get_column_count() < params.get_component_count()) {
+            throw invalid_argument("Data column count should be >= component count.");
+        }
```
Comment (PetrovKP) [inline cpp/oneapi/dal/algo/pca/detail/train_ops.hpp:43]:
```c++
params.get_component_count() < 0 -> throw
```

File: cpp/oneapi/dal/algo/kmeans/backend/cpu/train_kernel_lloyd_dense.cpp [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    daal_kmeans::Parameter par(cluster_count, max_iteration_count);
+    par.accuracyThreshold = accuracy_threshold;
+
+    auto arr_data        = row_accessor<const Float>{ data }.pull();
+    const auto daal_data = interop::convert_to_daal_homogen_table(arr_data,
+                                                                  data.get_row_count(),
+                                                                  data.get_column_count());
+
+    auto new_initial_centroids = initial_centroids;
+    if (!new_initial_centroids.has_data()) {
```
Comment (PetrovKP) [inline cpp/oneapi/dal/algo/kmeans/backend/cpu/train_kernel_lloyd_dense.cpp:63]: Is this planned to be done as an algorithm inside an algorithm?

File: cpp/oneapi/dal/algo/kmeans_init/common.cpp [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/algo/kmeans_init/common.hpp"
+
+namespace oneapi::dal::kmeans_init {
+
+class detail::descriptor_impl : public base {
+public:
+    std::int64_t cluster_count = -1;
```
Comment (PetrovKP) [inline cpp/oneapi/dal/algo/kmeans_init/common.cpp:23]: The default value is -1? What is it like?

File: cpp/oneapi/dal/algo/kmeans/backend/cpu/train_kernel_lloyd_dense.cpp (hunk 10 above)
Comment (mchernov-intel) [reply cpp/oneapi/dal/algo/kmeans/backend/cpu/train_kernel_lloyd_dense.cpp:63]: Decided to postpone

File: cpp/oneapi/dal/algo/kmeans/detail/train_ops.hpp [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    using descriptor_base_t = descriptor_base;
+
+    void check_preconditions(const Descriptor& params, const train_input& input) const {
+        if (!(input.get_data().has_data())) {
+            throw domain_error("Input data is empty.");
+        }
+        if (input.get_initial_centroids().has_data()) {
+            if (input.get_initial_centroids().get_row_count() != params.get_cluster_count()) {
+                throw invalid_argument(
+                    "Initial centroids row count is not equal to cluster count.");
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/kmeans/detail/train_ops.hpp:46]:
```suggestion
                    "Input initial_centroids row count should be equal to descriptor cluster_count");
```

File: cpp/oneapi/dal/algo/kmeans/detail/train_ops.hpp [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+            }
+            if (input.get_initial_centroids().get_column_count() !=
+                input.get_data().get_column_count()) {
+                throw invalid_argument(
+                    "Initial centroids column count is not equal data column count.");
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/kmeans/detail/train_ops.hpp:51]:
```suggestion
                    "Input initial_centroids column count should be equal to input data column count.");
```

File: cpp/oneapi/dal/algo/kmeans/detail/train_ops.hpp [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+            }
+        }
+    }
+
+    void check_postconditions(const Descriptor& params,
+                              const train_input& input,
+                              const train_result& result) const {
+        if (!(result.get_labels().has_data())) {
+            throw internal_error("Labels are empty.");
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/kmeans/detail/train_ops.hpp:60]:
```suggestion
            throw internal_error("Result labels should not be empty");
```

File: cpp/oneapi/dal/algo/kmeans/detail/train_ops.hpp [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+    void check_postconditions(const Descriptor& params,
+                              const train_input& input,
+                              const train_result& result) const {
+        if (!(result.get_labels().has_data())) {
+            throw internal_error("Result labels should not be empty");
+        }
+        if (result.get_labels().get_column_count() != input.get_data().get_row_count()) {
+            throw internal_error("Result labels column count should be equal to data row count");
+        }
+        if (!(result.get_model().get_centroids().has_data())) {
```
[the remark this answers is not in the document]
Comment (mchernov-intel) [reply cpp/oneapi/dal/algo/kmeans/detail/train_ops.hpp:63]: The result doesn't contain the centroids.

File: cpp/oneapi/dal/algo/kmeans/detail/infer_ops.hpp [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+            input.get_data().get_column_count()) {
+            throw invalid_argument(
+                "Model centroids column_count should be equal to input data column_count");
+        }
+    }
+
+    void check_postconditions(const Descriptor& params,
+                              const infer_input& input,
+                              const infer_result& result) const {
+        if (!(result.get_labels().has_data())) {
```
[the remark this answers is not in the document]
Comment (mchernov-intel) [reply cpp/oneapi/dal/algo/kmeans/detail/infer_ops.hpp:58]: TBD in the future

**Pull request merged**